File: leetcode/0828-count-unique-characters-of-all-substrings-of-a-given-string/uniqueLetterString.ts
function uniqueLetterString(s: string): number {
  const letterIndexesMap = new Map<string, number[]>();
  const stringLength = s.length;
  let result = 0;

  for (let index = 0; index < stringLength; index++) {
    const letter = s[index]!;
    const letterIndexes = letterIndexesMap.get(letter) ?? [];

    if (letterIndexes.length < 1) {
      letterIndexesMap.set(letter, letterIndexes);
    }

    letterIndexes.push(index);
  }

  for (const [, letterIndexes] of letterIndexesMap) {
    for (const [index, current] of letterIndexes.entries()) {
      const start = (letterIndexes[index - 1] ?? -1) + 1;
      const end = (letterIndexes[index + 1] ?? stringLength) - 1;

      result = result + (current - start + 1) * (end - current + 1);
    }
  }
  return result;
}

export default uniqueLetterString;

File: leetcode/0828-count-unique-characters-of-all-substrings-of-a-given-string/uniqueLetterString.go
package leetcode

func uniqueLetterString(s string) int {
	var letterIndexesMap = make(map[rune][]int)
	var result = 0
	var stringLength = len(s)

	for index, letter := range s {
		letterIndexesMap[letter] = append(letterIndexesMap[letter], index)
	}

	for _, letterIndexes := range letterIndexesMap {
		for index, current := range letterIndexes {
			var start int
			var end int
			if index == 0 {
				start = 0
			} else {
				start = letterIndexes[index-1] + 1
			}
			if index == len(letterIndexes)-1 {
				end = stringLength - 1
			} else {
				end = letterIndexes[index+1] - 1
			}

			result = result + (current-start+1)*(end-current+1)
		}
	}

	return result
}
